src/hydra-queue-runner/queue-monitor.cc

```diff
@@ -74,32 +70,27 @@ struct PreviousFailure : public std::exception {
 };
 
 
-bool State::getQueuedBuilds(Connection & conn,
-    ref<Store> destStore, unsigned int & lastBuildId)
+bool State::getQueuedBuilds(Connection & conn, ref<Store> destStore)
 {
-    printInfo("checking the queue for builds > %d...", lastBuildId);
+    printInfo("checking the queue for new builds...");
 
     /* Grab the queued builds from the database, but don't process
        them yet (since we don't want a long-running transaction). */
     std::vector<BuildID> newIDs;
     std::map<BuildID, Build::ptr> newBuildsByID;
     std::multimap<StorePath, BuildID> newBuildsByPath;
 
-    unsigned int newLastBuildId = lastBuildId;
-
     {
         pqxx::work txn(conn);
 
         auto res = txn.exec_params
             ("select id, project, jobset, job, drvPath, maxsilent, timeout, timestamp, globalPriority, priority from Builds "
-             "where id > $1 and finished = 0 order by globalPriority desc, id",
-            lastBuildId);
+             "where finished = 0 order by globalPriority desc, id");
```

Comment:
Doesn't this cause it to select the same builds over and over (at least until they're finished)?

Reply:
Thanks for reviewing this, @edolstra.

As I understood the code, this particular section is building the `newIDs` list that is then processed (by the last `for` loop in this same function) to call `createBuild` which will then try `createStep` for the build's `drvPath` but the latter will find the step already existing in the `buildsteps` table and simply return that without adding to `newSteps`.  The result of that will be a call to `propagatePriorities` and a message, which I observe periodically in the logs:

    Jun 05 00:01:40 ___ hydra-queue-runner[14191]: added build 3836338 (top-level step /nix/store/___.drv, 0 new steps)

I suppose the `propagatePriorities` call and the log message could be skipped if `newSteps.empty()`.  I can make this change if you agree with this tactic.

Ultimately yes, this causes the queue runner to do some extra work re-checking builds that are already queued until they finish, but I think that cost is relatively low compared to leaving stranded pending builds that never get processed.  More efficiency could probably be gained by adding another state to to the Builds table (perhaps to `buildStatus`), but that would be a considerably more impactful change which I tried to avoid in this PR.

Reply:
On hydra.nixos.org, we can have ~100000 unfinished builds in the database, so that means we would be constantly fetching the same 100000 builds from the database (and processing them). That's a lot of database traffic...